exchanges/account/account.go

```diff
@@ -39,80 +52,255 @@ func CollectBalances(accountBalances map[string][]Balance, assetType asset.Item)
+				continue
+			}
+
+			accountsHoldings = append(accountsHoldings, SubAccount{
+				ID:         subAccount,
+				AssetType:  ai,
+				Currencies: currencyBalances,
+			})
+			break
+		}
+	}
+
+	if len(accountsHoldings) == 0 {
+		return Holdings{}, fmt.Errorf("%s %s %w",
+			exch,
+			assetType,
+			errAssetHoldingsNotFound)
+	}
+	return Holdings{Exchange: exch, Accounts: accountsHoldings}, nil
 }
 
-// GetHoldings returns full holdings for an exchange
-func GetHoldings(exch string, assetType asset.Item) (Holdings, error) {
+// GetBalance returns the internal balance for that asset item.
+func GetBalance(exch, subAccount string, ai asset.Item, c currency.Code) (*ProtectedBalance, error) {
 	if exch == "" {
-		return Holdings{}, errors.New("exchange name unset")
+		return nil, errExchangeNameUnset
+	}
+
+	if !ai.IsValid() {
+		return nil, fmt.Errorf("%s %w", ai, asset.ErrNotSupported)
+	}
+
+	if c.IsEmpty() {
+		return nil, currency.ErrCurrencyCodeEmpty
 	}
 
 	exch = strings.ToLower(exch)
+	subAccount = strings.ToLower(subAccount)
 
-	if !assetType.IsValid() {
-		return Holdings{}, fmt.Errorf("assetType %v is invalid", assetType)
+	service.m.Lock()
+	defer service.m.Unlock()
+
+	accounts, ok := service.exchangeAccounts[exch]
+	if !ok {
+		return nil, fmt.Errorf("%s %w", exch, errExchangeHoldingsNotFound)
 	}
 
-	service.Lock()
-	defer service.Unlock()
-	h, ok := service.accounts[exch]
+	assetBalances, ok := accounts.SubAccounts[subAccount]
 	if !ok {
-		return Holdings{}, errors.New("exchange account holdings not found")
+		return nil, fmt.Errorf("%s %s %w",
+			exch, subAccount, errNoExchangeSubAccountBalances)
 	}
-	for y := range h.h.Accounts {
-		if h.h.Accounts[y].AssetType == assetType {
-			return *h.h, nil
-		}
+
+	currencyBalances, ok := assetBalances[ai]
+	if !ok {
+		return nil, fmt.Errorf("%s %s %s %w",
+			exch, subAccount, ai, errAssetHoldingsNotFound)
+	}
+
+	bal, ok := currencyBalances[c.Item]
+	if !ok {
+		return nil, fmt.Errorf("%s %s %s %s %w",
+			exch, subAccount, ai, c, errNoBalanceFound)
 	}
-	return Holdings{}, fmt.Errorf("%v holdings data not found for %s", assetType, exch)
+	return bal, nil
 }
 
 // Update updates holdings with new account info
 func (s *Service) Update(a *Holdings) error {
+	if a == nil {
+		return errHoldingsIsNil
+	}
+
+	if a.Exchange == "" {
+		return errExchangeNameUnset
+	}
+
 	exch := strings.ToLower(a.Exchange)
-	s.Lock()
-	acc, ok := s.accounts[exch]
+	s.m.Lock()
+	defer s.m.Unlock()
+	accounts, ok := s.exchangeAccounts[exch]
 	if !ok {
 		id, err := s.mux.GetID()
 		if err != nil {
-			s.Unlock()
 			return err
 		}
+		accounts = &Accounts{
+			ID:          id,
+			SubAccounts: make(map[string]map[asset.Item]map[*currency.Item]*ProtectedBalance),
+		}
+		s.exchangeAccounts[exch] = accounts
+	}
+
+	var errs common.Errors
+	for x := range a.Accounts {
+		if !a.Accounts[x].AssetType.IsValid() {
+			errs = append(errs, fmt.Errorf("cannot load sub account holdings for %s [%s] %w",
```

Comment:
This change has some big ramifications for GCT. Any exchange implementation of `UpdateAccountInfo` where an `asset.Item` is not set, will return this error and then disable all authenticated use for that exchange in GCT
```shell
[WARN]  | EXCHANGE |  14/04/2022 09:35:31  | Bitfinex: Cannot validate credentials, authenticated support has 
been disabled, Error: cannot load sub account holdings for deposit [] received unsupported asset type, cannot 
load sub account holdings for exchange [] received unsupported asset type, cannot load sub account holdings fo
r trading [] received unsupported asset type, cannot load sub account holdings for margin [] received unsupported asset type, cannot load sub account holdings for funding  [] received unsupported asset type
```
This is a regression and needs to be fixed before merge